### ISSUE TITLE
Update S4457: FP for async void method (event handlers)

### ISSUE THREAD

**Issue by freza-tm** [SonarSource/sonar-dotnet#2665]
### Description

There is no way of splitting async void method into two without breaking another sonar rules. 

### Repro steps

```
private async void OnButtonPressed( object sender, EventArgs e)
{
  if(e == null)
    throw new ArgumentException(nameof(e));

  await Task.Delay(0);
}
```

### Expected behavior

the S4457 should not be raised for async void methods

### Actual behavior

S4457 is raised

### Related information

* SonarLint for VS2017 4.12.0.0
* VS 2017

**Comment by pavel-mikula-sonarsource** [comment]
Hi @freza-tm,

Can you please provide your attempts to split the code and the list of other sonar rules that it breaks? These two ways seems to work fine:

```
private void OnButtonPressed(object sender, EventArgs e)
{
    if (e == null)
        throw new ArgumentException(nameof(e));

    Task.Run(async () => await OnButtonPressedInternal());
}

private async Task OnButtonPressedInternal()
{
    await Task.Delay(0);
}
```

**Comment by freza-tm** [comment]
Hi @pavel-mikula-sonarsource 
I think the solution you wrote is not correnct nor safe. The OnButtonPressedInternal will be executed on threadpool and not the UI thread which can lead to serious problems (cross thread access). Also any exception thrown inside of OnButtonPressedInternal will be unhandled and terminate the process.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @freza-tm ,
Discussion about how async/await and try/catch itself should be treated is out of scope of this FP report. 
We'll need more information about how you've tried to refactor the rule and what other rules were violated to confirm this as FP.

**Comment by freza-tm** [comment]
Solution proposed in the example (https://rules.sonarsource.com/csharp/RSPEC-4457) is to extract await code after it to another method, which in case of async event handler is not possible without creating void async method and that generates S3168. If I change the return type to Task then CS4014 (async method not awaited).

Only solution that remove the S4457 is to extract the validation to external method, but that seems to defy the point of this warning

**Comment by StingyJack** [comment]
@andrei-epure-sonarsource  will the fix address S4457 being raised for console applications?
```
private static async Task Main(string[] args)
{
 ...
```

Trying to make this use a `void` return instead of a task will just make the method signature get flagged for S3168

**Comment by andrei-epure-sonarsource** [comment]
Hi @StingyJack , sorry, I saw now your comment. This issue, as the title suggests, fixes FPs for async void methods.

Please open an issue detailing the FP.
